File: WatchApp/app/src/main/java/com/speakbyhand/app/core/GestureDetector.java
package com.speakbyhand.app.core;

import android.content.Context;

import com.speakbyhand.app.ml.GestureConvModel;

import org.tensorflow.lite.DataType;
import org.tensorflow.lite.support.tensorbuffer.TensorBuffer;

import java.io.IOException;

public class GestureDetector {
    Context context;


    public GestureCode detect(GestureData data){
        try {
            GestureConvModel model = GestureConvModel.newInstance(context);

            TensorBuffer inputFeatures = TensorBuffer.createFixedSize(new int[]{1, 197, 6}, DataType.FLOAT32);
            inputFeatures.loadArray(data.toArray());

            GestureConvModel.Outputs outputs = model.process(inputFeatures);
            TensorBuffer outputFeatures = outputs.getOutputFeature0AsTensorBuffer();
            model.close();

            int predictionIndex = getArgMax(outputFeatures.getFloatArray());
            return toGestureCode(predictionIndex);
        } catch (IOException e) {
            // TODO Handle the exception
        }

        return GestureCode.DrinkWater;
    }

    // TODO: Create unit test for this method
    private int getArgMax(float[] floatArray) {
        int maxIndex = 0;
        for (int i = 0; i < floatArray.length; i++) {
            if(floatArray[i] > floatArray[maxIndex]){
                maxIndex = i;
            }
        }
        return maxIndex;
    }

    // TODO: Create unit test for this method
    private GestureCode toGestureCode(int predictionIndex){
        switch (predictionIndex){
            case 0:
                return GestureCode.DrinkWater;
            case 1:
                return GestureCode.EatFood;
            case 2:
                return GestureCode.Help;
            case 3:
                return GestureCode.No;
            case 4:
                return GestureCode.Toilet;
            case 5:
                return GestureCode.Yes;
        }

        throw new IllegalArgumentException("Given argument: "+predictionIndex);
    }
}

File: WatchApp/app/src/main/java/com/speakbyhand/app/core/GestureDetector.kt
package com.speakbyhand.app.core

import android.content.Context
import android.content.res.AssetFileDescriptor
import android.util.Log
import org.tensorflow.lite.DataType
import org.tensorflow.lite.Interpreter
import org.tensorflow.lite.support.tensorbuffer.TensorBuffer
import java.io.FileInputStream
import java.nio.ByteBuffer
import java.nio.channels.FileChannel


class GestureDetector(var context: Context, private val model_file_name: String) {

    fun detect(data: GestureData): GestureCode {
        val interpreter = Interpreter(readModelFile())

        val input = TensorBuffer.createFixedSize(intArrayOf(1, 297, 6), DataType.FLOAT32)
        val output = TensorBuffer.createFixedSize(intArrayOf(1, 6), DataType.FLOAT32)
        input.loadArray(data.toArray())

        interpreter.run(input.buffer, output.buffer)

        return if(arePredictionsAboveConfidenceThreshold(output.floatArray)){
            val predictionIndex = getArgMax(output.floatArray)
            toGestureCode(predictionIndex)
        } else {
            GestureCode.Unknown
        }
    }

    private fun readModelFile() : ByteBuffer {
        val fileDescriptor: AssetFileDescriptor = context.assets.openFd(model_file_name)
        val inputStream = FileInputStream(fileDescriptor.fileDescriptor)
        val fileChannel: FileChannel = inputStream.channel
        val startOffset = fileDescriptor.startOffset
        val declareLength = fileDescriptor.declaredLength
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declareLength)
    }

    private fun arePredictionsAboveConfidenceThreshold(floatArray: FloatArray) : Boolean{
        for (i in floatArray.indices) {
            val confidenceThreshold = .7f
            if (floatArray[i] > confidenceThreshold) {
                return true
            }
        }
        return false
    }

    private fun getArgMax(floatArray: FloatArray): Int {
        var maxIndex = 0
        for (i in floatArray.indices) {
            if (floatArray[i] > floatArray[maxIndex]) {
                maxIndex = i
            }
        }
        return maxIndex
    }

    private fun toGestureCode(predictionIndex: Int): GestureCode {
        when (predictionIndex) {
            0 -> return GestureCode.DrinkWater
            1 -> return GestureCode.EatFood
            2 -> return GestureCode.Help
            3 -> return GestureCode.No
            4 -> return GestureCode.Toilet
            5 -> return GestureCode.Yes
        }
        throw IllegalArgumentException("Given argument: $predictionIndex")
    }
}